src/Maestro/Maestro.Web/Startup.cs

```diff
@@ -43,16 +47,40 @@ public partial class Startup
     {
         static Startup()
         {
-            Triggers<BuildChannel>.Inserted += entry =>
+            Triggers<BuildChannel>.Inserting += entry =>
             {
+                BuildAssetRegistryContext context = entry.Context as BuildAssetRegistryContext;
                 BuildChannel entity = entry.Entity;
+                if (ChannelHasAssociatedReleasePipeline(entity.ChannelId, context))
+                {
+                    entry.Cancel = true;
```

Comment:
>entry.Cancel = true; [](start = 20, length = 20)

Cancel? Does this stop it from inserting into the DB??? What does that do to the caller?

Reply:
Yes, it stops it from being set in the DB. The caller still sees a 201.

Reply:
This is... super hacky.  We shouldn't be using a DB insert to trigger a build and not actually insert into the database. It's extremely unexpected for something like `await _context.BuildChannels.AddAsync(buildChannel);` doesn't actually insert into the database. And it means the web call will say "I added it to the channel", but then if you query that build, it won't be in that channel.

Really the callers that are calling "add to channel" shouldn't be doing that. We should fix the callers so that they call "DoAppropriateReleaseStuffAsync" instead, and that does the right thing.

Having a trigger you can't see in the code that cancels the insert and lies to a web request that schedules a magic release that implicitly will fix... whatever... is way too much implicit magic and is a maintenance nightmare.  It took Ricardo and I almost 15 minutes just to find where we _think_ this code was being triggered from, and we're only like 50% sure that's actually the right place.  Triggers are *_BAD_*.

---
In reply to: [260462310](https://github.com/dotnet/arcade-services/pull/125#discussion_r260462310) [](ancestors = 260462310)

Reply:
Chatted offline.

It's a little late to change the design at this point. Definitely something to consider as a future improvement.